lib/DxilValidation/DxilContainerValidation.cpp

```diff
@@ -73,6 +78,522 @@ struct DiagRestore {
+        ShaderStage == PSVShaderKind::Geometry ? PSV_GS_MAX_STREAMS : 1;
+
+    if (UsesViewID) {
+      for (unsigned i = 0; i < NumStreams; ++i) {
+        OS << "Outputs affected by ViewID as a bitmask for stream " << i
+           << ":\n ";
+        uint8_t OutputVectors =
+            llvm::RoundUpToAlignment(NumOutputSigScalars[i], 4) / 4;
+        const PSVComponentMask ViewIDMask(ViewIDOutputMask[i].data(),
+                                          OutputVectors);
+        std::string OutputSetName = "Outputs";
+        OutputSetName += "[" + std::to_string(i) + "]";
+        ViewIDMask.Print(OS, "ViewID", OutputSetName.c_str());
+      }
+
+      if (ShaderStage == PSVShaderKind::Hull ||
+          ShaderStage == PSVShaderKind::Mesh) {
+        OS << "PCOutputs affected by ViewID as a bitmask:\n";
+        uint8_t OutputVectors =
+            llvm::RoundUpToAlignment(NumPCOrPrimSigScalars, 4) / 4;
+        const PSVComponentMask ViewIDMask(ViewIDPCOutputMask.data(),
+                                          OutputVectors);
+        ViewIDMask.Print(OS, "ViewID", "PCOutputs");
+      }
+    }
+
+    for (unsigned i = 0; i < NumStreams; ++i) {
+      OS << "Outputs affected by inputs as a table of bitmasks for stream " << i
+         << ":\n";
+      uint8_t InputVectors =
+          llvm::RoundUpToAlignment(NumInputSigScalars, 4) / 4;
+      uint8_t OutputVectors =
+          llvm::RoundUpToAlignment(NumOutputSigScalars[i], 4) / 4;
+      const PSVDependencyTable Table(InputToOutputTable[i].data(), InputVectors,
+                                     OutputVectors);
+      std::string OutputSetName = "Outputs";
+      OutputSetName += "[" + std::to_string(i) + "]";
+      Table.Print(OS, "Inputs", OutputSetName.c_str());
+    }
+
+    if (ShaderStage == PSVShaderKind::Hull ||
+        ShaderStage == PSVShaderKind::Mesh) {
+      OS << "Patch constant outputs affected by inputs as a table of "
+            "bitmasks:\n";
+      uint8_t InputVectors =
+          llvm::RoundUpToAlignment(NumInputSigScalars, 4) / 4;
+      uint8_t OutputVectors =
+          llvm::RoundUpToAlignment(NumPCOrPrimSigScalars, 4) / 4;
+      const PSVDependencyTable Table(InputToPCOutputTable.data(), InputVectors,
+                                     OutputVectors);
+      Table.Print(OS, "Inputs", "PatchConstantOutputs");
+    } else if (ShaderStage == PSVShaderKind::Domain) {
+      OS << "Outputs affected by patch constant inputs as a table of "
+            "bitmasks:\n";
+      uint8_t InputVectors =
+          llvm::RoundUpToAlignment(NumPCOrPrimSigScalars, 4) / 4;
+      uint8_t OutputVectors =
+          llvm::RoundUpToAlignment(NumOutputSigScalars[0], 4) / 4;
+      const PSVDependencyTable Table(PCInputToOutputTable.data(), InputVectors,
+                                     OutputVectors);
+      Table.Print(OS, "PatchConstantInputs", "Outputs");
+    }
+  }
+};
+SimpleViewIDState::SimpleViewIDState(std::vector<uint32_t> &Data,
```

Comment:
We now have largely the same code in three places: `DxilViewIdState::Serialize`, `ComputeSerializedViewIDStateSizeInUInts` (which suggests merging itself with the first) and here. Can we not collect this somehow? If nothing else, we should backport the switch to `llvm::RoundUpToAlignment` to those locations.